src/main/scala/me/jeffshaw/digitalocean/Action.scala

```diff
@@ -29,22 +31,25 @@ case class Action(
    * @return
    */
   def complete()(implicit client: DigitalOceanClient, ec: ExecutionContext): Future[Action] = {
-    def actionRefresh: Action = {
-      Thread.sleep(client.actionCheckInterval.toMillis)
-      Await.result(Action(id), client.maxWaitPerRequest)
-    }
 
-    val actionRefreshPoll = Iterator.continually(actionRefresh)
+    import DelayedFuture._
 
-    for {
-      completedAction <- Future(actionRefreshPoll.find(_.isCompleted).get)
-    } yield {
-      if (completedAction.status == Action.Errored) {
-        throw new ActionErroredException(completedAction)
-      } else {
-        completedAction
+    def loop(): Future[Action] = {
+      val timeout = client.actionCheckInterval + client.maxWaitPerRequest
+      val whenAction = after(client.actionCheckInterval)(Action(id))
+      val whenTimeout = after(timeout)(Future.failed(new TimeoutException()))
+      val whenActionWithoutTimeout = Future.firstCompletedOf(Seq(whenAction, whenTimeout))
+
+      whenActionWithoutTimeout.flatMap { action =>
+        action.status match {
+          case InProgress => loop()
```

Comment:
When the new status is InProgress, there needs to be a delay of "client.actionCheckInterval" before "loop()". Otherwise the client is in more danger of hitting its request rate limit.